[lldb] Implement CLI support for reverse-continue

rocallahan opened the pull request:
This introduces the options "-F/--forward" and "-R/--reverse" to `process continue`.

These only work if you're running with a gdbserver backend that supports reverse execution, such as rr. For testing we rely on the fake reverse-execution functionality in `lldbreverse.py`.

rocallahan committed:
[lldb] Implement CLI support for reverse-continue

This introduces the options "-F/--forward" and
"-R/--reverse" to `process continue`.

These only work if you're running with a gdbserver
backend that supports reverse execution, such as
rr. For testing we rely on the fake reverse-
execution functionality in `lldbreverse.py`.

rocallahan commented:
I suppose we need to add documentation for this but I'm not sure where...

vogelsgesang commented:
> I suppose we need to add documentation for this but I'm not sure where...

Afaict, your code change should already be sufficient such that `help processor continue` also lists the new -F and -R flags.

In addition, I guess we would eventually want to also document more generally how to use rr with lldb (how to start the rr replay, how to connect lldb to it, etc.) such that end users are able to discover this feature. However, I guess before doing so, we would like to also implement reverse single- stepping etc.?

vogelsgesang commented on lldb/test/API/commands/process/reverse-continue/TestReverseContinue.py:
We probably also want to test that
* reverse-continue on a non-reverse-capable process gives a decent error message
* `process continue -F` also work for non-reverse-capable processes

vogelsgesang commented on lldb/source/Commands/CommandObjectProcess.cpp:
```diff
@@ -468,7 +468,23 @@ class CommandObjectProcessContinue : public CommandObjectParsed {
+      case 'R':
+        if (m_base_direction == lldb::RunDirection::eRunForward) {
+          error = Status::FromErrorString(
+              "cannot specify both 'forward' and 'reverse'");
```
We might want to add a test case for this error condition?

rocallahan commented:
> In addition, I guess we would eventually want to also document more generally how to use rr with lldb (how to start the rr replay, how to connect lldb to it, etc.) such that end users are able to discover this feature. However, I guess before doing so, we would like to also implement reverse single- stepping etc.?

Yes, but actually I think this documentation can be in rr.

Reverse-continue is 80% of what people do in practice so just advertising "lldb supports reverse-continue with rr now" would be OK.

> reverse-continue on a non-reverse-capable process gives a decent error message

Added.

> process continue --forward also work for non-reverse-capable processes

Added.

> We might want to add a test case for this error condition?

Added.

JDevlieghere commented on lldb/source/Commands/CommandObjectProcess.cpp:
```diff
@@ -654,6 +672,10 @@ class CommandObjectProcessContinue : public CommandObjectParsed {
         }
       }
 
+      if (m_options.m_base_direction.has_value()) {
+        process->SetBaseDirection(*m_options.m_base_direction);
+      }
```
```suggestion
      if (m_options.m_base_direction.has_value())
        process->SetBaseDirection(*m_options.m_base_direction);
```

vogelsgesang commented:
> Yes, but actually I think this documentation can be in rr.

Agree, we should probably have the bulk of the documentation in the rr repo. Not sure if we want to put it online right away, or if we wait to wait for lldb-21 to go out the door first. (Not sure how accustomed rr users are to building lldb from source)

Inside the LLVM project, we probably want to add a release note for this change, in the lldb section in `llvm/docs/ReleaseNotes.md`. Would be great I'd you could add a short bullet point there as part of this PR

rocallahan commented:
> or if we wait to wait for lldb-21 to go out the door first

I'll definitely wait for the next lldb release before updating the rr docs.

> Would be great I'd you could add a short bullet point there

Done.

medismailben commented on lldb/source/Commands/Options.td:
```diff
@@ -744,6 +744,10 @@ let Command = "process continue" in {
     Arg<"BreakpointIDRange">, Desc<"Specify a breakpoint to continue to, temporarily "
     "ignoring other breakpoints.  Can be specified more than once.  "
     "The continue action will be done synchronously if this option is specified.">;
+  def thread_continue_forward : Option<"forward", "F">, Group<3>,
```
I don't think it's necessary to have 2 options for this, I'd just keep the reverse option and make it default to false. That would simplify the logic in the CommandObject as well.

rocallahan replied:
Processes have a persistent base direction. `--forward` forces forward, `--reverse` forces reverse, and neither means continue in the current dirction.

jimingham commented on lldb/source/Commands/CommandObjectProcess.cpp:
```diff
@@ -468,7 +468,23 @@ class CommandObjectProcessContinue : public CommandObjectParsed {
         m_run_to_bkpt_args.AppendArgument(option_arg);
         m_any_bkpts_specified = true;
-      break;
+        break;
```
This isn't how you encode two incompatible options in lldb's command parser.  Just put the two options in two Option Sets and lldb will take care of the rest.  Look for uses of LLDB_OPT_SET for more details.

rocallahan commented on lldb/source/Commands/Options.td:
```diff
@@ -744,6 +744,10 @@ let Command = "process continue" in {
+  def thread_continue_forward : Option<"forward", "F">, Group<3>,
+    Desc<"Execute in forward direction">;
+  def thread_continue_reverse : Option<"reverse", "R">, Group<3>,
+    Desc<"Execute in reverse direction">;
```
Thanks. I'm using four groups.

DavidSpickett commented on llvm/docs/ReleaseNotes.md:
```diff
@@ -220,6 +220,8 @@ Changes to LLDB
   information about the current state of the debugger at the bottom of the
   terminal. This is on by default and can be configured using the
   `show-statusline` and `statusline-format` settings.
+* LLDB now supports `process continue --reverse` when used with backends
```
`--forward` is also new, mention that. Maybe:
"When using reverse execution, `process continue --forward` can be used to return to forward execution."

Then readers have all they need to try this out without going into `help process continue`.

rocallahan replied:
> backends -> "debug servers"

Done.

> --forward is also new, mention that

Done. There is more that could be said, e.g. how stepping commands will currently reset the direction to forward, but we can't put everything here.

DavidSpickett replied:
Agreed. Reader's have --reverse and --forward, that's enough to experiment with.

rocallahan commented:
Thanks for the feedback, folks... if you're happy with it, don't forget to give it the thumbs-up :-)

rocallahan commented:
@jimingham @DavidSpickett can one of you approve this please?

JDevlieghere reviewed (approved):
LGTM

rocallahan commented:
Thanks. Can this be merged now?

luporl commented:
It seems this broke lldb-aarch64-windows bot: https://lab.llvm.org/buildbot/#/builders/141/builds/8149

DavidSpickett commented:
Likely won't find the fix today, but I suspect that the Windows process class just doesn't know that it doesn't support reverse execution:
```
(lldb) process continue --reverse
Process 3108 resuming
(lldb) process status
Process 5360 stopped
```
The process being stopped is probably due to us failing to continue properly.

labath commented:
Probably because windows uses a different (in-proces) implementation.
